File: python/parse_json_with_keys.py
"""
This code fetch the values of a given key from a nested JSON file recursively.
Input: JSON file and a Key
Output: List of values for the Key
"""

import json


def json_extract(json_object, key):
    """Recursively fetch values from nested JSON."""
    values_list = []

    def extract(obj, value, key):
        """Recursively search for values of key in JSON tree."""
        if isinstance(obj, dict):
            for k, v in obj.items():
                if isinstance(v, (dict, list)):
                    extract(v, value, key)
                elif k == key:
                    value.append(v)

        elif isinstance(obj, list):
            for item in obj:
                extract(item, value, key)
        return value

    values = extract(json_object, values_list, key)
    return values


if __name__ == '__main__':
    json_file_path = '../data/example.json'
    with open(json_file_path, 'r') as json_file:
        json_file_data = json.load(json_file)
        values_for_given_key = json_extract(json_file_data, "text")
        print(values_for_given_key.__len__())
        print(values_for_given_key)

File: src/main/java/ParseJSONWithKeys.java
/*
This code fetch the values of a given key from a nested JSON file recursively.
Input: JSON file and a Key
Output: Values for the Key
*/
import org.json.JSONArray;
import org.json.simple.parser.JSONParser;
import org.json.JSONObject;
import java.io.FileReader;
import java.util.Iterator;

public class ParseJSONWithKeys
{
    public static void getValues(JSONObject json, String key)
    {
        boolean exists = json.has(key);
        Iterator<?> keys;
        String nextKeys;

        if(!exists)
        {
            keys = json.keys();
            while(keys.hasNext())
            {
                nextKeys = (String) keys.next();
                try
                {
                    if(json.get(nextKeys) instanceof JSONObject)
                    {
                        getValues(json.getJSONObject(nextKeys), key);
                    }
                    else if(json.get(nextKeys) instanceof JSONArray)
                    {
                        JSONArray jsonArray = json.getJSONArray(nextKeys);
                        for (int i = 0; i < jsonArray.length(); i++)
                        {
                            String jsonArrayString= jsonArray.get(i).toString();
                            JSONObject innerJSON = new JSONObject(jsonArrayString);

                            getValues(innerJSON, key);
                        }
                    }
                }
                catch (Exception e)
                {

                }
            }
        }
        else
        {
            System.out.println(json.get(key));
        }
    }

    public static void main(String [] args)
    {
        String jsonFilePath = "data/example.json";
        JSONParser jsonParser = new JSONParser();
        JSONObject inputJSONObject;

        try (FileReader reader = new FileReader(jsonFilePath))
        {
            //Read JSON file
            String obj = jsonParser.parse(reader).toString();
            inputJSONObject = new JSONObject(obj);

            getValues(inputJSONObject, "text");
        } catch (Exception e)
        {
            e.printStackTrace();
        }
    }
}